Added implementation for calculating number of lines added and removed

maha-sachin opened the pull request:
Added modification_counter function in interactive_viewer.ml that will provide added and removed line count, also updated formatting as per contribution document.
fix for issue: https://github.com/panglesd/diffcessible/issues/17
Sample output: 
<img width="745" alt="Screenshot 2024-03-06 at 7 46 57 PM" src="https://github.com/panglesd/diffcessible/assets/122064139/56cb3a3d-f762-4ba2-9d8a-651dac83b8ba">

@Julow and @panglesd

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -52,20 +46,47 @@ let current_operation patches =
+  | Some p ->
+      let hunks = p.Patch.hunks in
+      let additions, removals =
+        List.fold_left
+          (fun (add_acc, remove_acc) hunk ->
+            let addition, removal = calculate_add_remove hunk.Patch.lines in
+            (add_acc + addition, remove_acc + removal))
+          (0, 0) hunks
```
Suggestion: move this and `calculate_add_remove` (renamed to `calculate_add_remove_hunk`) into a function named `calculate_add_remove_operation`. It makes it clearer what function does what!

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -52,20 +46,47 @@ let current_operation patches =
+      match line with
+      | `Their _ -> (add + 1, rem)
+      | `Mine _ -> (add, rem + 1)
+      | _ -> (add, rem))
```
It is good practice to try to have exhaustive match (without wildcards) as much as possible. This helps detect when there is a breaking change: for instance, the patch library add a 4th constructor. This 4th constructor would be included in the  `_` joker case, and the compiler would not help us detect that there is a change and that we need to handle it. With an "exhaustive" match, the addition of the 4th constructor would raise a "non-exhaustive match" warning and help us detect the change, so we can handle it!

panglesd reviewed (approved):
Nicely done!

I have left some comments in the code that require small changes, but it is very good in general.

panglesd commented:
Oh, and two more comments:

- It would be nice to handle the "plural": "1 addition", but "3 additions", and similarly "1 removal" but "5 removals"!
- You will need to rebase your PR on top of master since there is some unfortunate PR (due to ocamlformat...)

maha-sachin commented:
I have addressed all the above comments and had a little trouble with the latest code hence the delay. 
Do let me know if you have more comments. 

Also, the updated UI screenshot of the implementations is below 
<img width="991" alt="Screenshot 2024-03-07 at 11 44 47 AM" src="https://github.com/panglesd/diffcessible/assets/122064139/1be0be83-28e2-4245-821d-75fb07e9005a">

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -60,9 +60,38 @@ let navigate z_patches (dir : direction) : unit =
-(* let pure_str s = Lwd.pure (W.string s) *)
 let quit = Lwd.var false
 
+let calculate_add_remove_operation z_patches : ui Lwd.t =
```
Sorry if I wasn't clear! The organization in functions is not yet the one I had in mind.

We need to separate two distinct things:
- The computation of the two integers corresponding to the addition and removal
- and the creation of a ui element.

Ideally, the code for both should be in sufficiently distinct place, for instance two different function (or one function in another, but with a naming that suggests what it does.

But here, the `calculate_add_remove_operation` does not seem to return a UI element!

maha-sachin replied:
@panglesd, Thank you for the clarification! I understand now. I'll work on separating the concerns into distinct functions:
1. One function will handle the computation of the integers corresponding to the addition and removal.
2. Another function will be responsible for creating the UI element based on those computed values.
I'll ensure that the code is organized in a way that reflects these separate responsibilities. Thank you for your patience!

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -60,9 +60,38 @@ let navigate z_patches (dir : direction) : unit =
+  let additions, removals =
+    List.fold_left
+      (fun (add_acc, remove_acc) hunk ->
+        let calculate_add_remove_hunk lines =
```
Including the function definition inside the `fold_left` is less readable and "redefines it" each time, in theory.

maha-sachin commented:
Hi @panglesd, let me know what you think.

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -72,6 +109,7 @@ let view (patches : Patch.t list) =
   W.vbox
     [
       operation_info z_patches;
+      calculate_addition_removal_operation z_patches;
       current_operation z_patches;
```
We have `operation_info`, `current_operation` and `current_hunks` here. The name `calculate_addition_removal_operation` is much longer and feels different: other names are descriptive of what they will display (they are ui components) while `calculate` seems to be an instruction. Try to make your name more similar to the one here! Ui component might need computation, but that's not how you describe them.

panglesd reviewed (commented):
Great, that's very close from being merged, but there are still some peculiar things!

- `count_lines_in_hunks` does not count the number of lines in a hunk. It take a _single_ line and an accumulator and adds the contribution of the line to the accumulator. So the name is very misleading!
- Similarly `calculate_addition_removal_in_patches` does not take multiple patches as input, but rather multiple `hunks`, so the name is also misleading.

Let me suggest some improvements!
- Define the functions that will never be useful outside of this context as local variables (for instance, the one you call `count_lines_in_hunk` should be local)
- Give the local functions simpler name. We already know the context since they are local, so they can be named, say, `add_hunk` (for instance) if they add the contribution of a hunk in the total number of additions and removal.
- Your code will look better with a `add_line` function.

maha-sachin commented:
@panglesd, Could you please take a look at the updates I made based on your suggestions? I've made several changes to improve the clarity and consistency of the function names. I would appreciate your feedback on the adjustments. Thank you

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -60,9 +60,40 @@ let navigate z_patches (dir : direction) : unit =
+  List.fold_left
+    (fun total_counts hunk ->
+      let additions, removals = add_lines hunk.Patch.lines in
+      (fun (add1, remove1) (add2, remove2) -> (add1 + add2, remove1 + remove2))
+        total_counts (additions, removals))
```
You are creating an anonymous function that you are directly applying to some arguments.

This example is more obvious but very similar to yours:
```
(fun x y -> x + y) 4 10
```
is a convoluted version of
```
4 + 10
```

Could you change this similarly to avoid the unnecessary function?

panglesd reviewed (commented):
That's better!
There are still some things I'd like to improve. Functions that "add" something should take an accumulator (the number of additions/removal) and add the contribution of a line/hunk to this accumulator.
So, for instance, the function defined [line 66](https://github.com/panglesd/diffcessible/pull/30/files#diff-0b2d0a78372bb1d9e5c8972b915d555fee362e1cbd95e9a1d48ddf9a2812b5cbR66) is an "add" function and its name is right.
However, the functions defined line 65 and 74 are _not_ an "add" functions. They could be turned into "add" functions by giving an accumulator to add the contributions to. I believe you do want to add the accumulator to the function line 65, but maybe only rename the function line 65, eg to `additions_and_removal`!

maha-sachin commented:
> That's better! There are still some things I'd like to improve. Functions that "add" something should take an accumulator (the number of additions/removal) and add the contribution of a line/hunk to this accumulator. So, for instance, the function defined [line 66](https://github.com/panglesd/diffcessible/pull/30/files#diff-0b2d0a78372bb1d9e5c8972b915d555fee362e1cbd95e9a1d48ddf9a2812b5cbR66) is an "add" function and its name is right. However, the functions defined line 65 and 74 are _not_ an "add" functions. They could be turned into "add" functions by giving an accumulator to add the contributions to. I believe you do want to add the accumulator to the function line 65, but maybe only rename the function line 65, eg to `additions_and_removal`!

@panglesd, Would you prefer to rename add_hunk to something like accumulate_count or combine_hunks_counts to better reflect its purpose?

maha-sachin commented:
@Julow & @panglesd, Could you please look at the updates I made based on your suggestions? Thank you and I would appreciate your feedback on the adjustments.

Julow commented on lib/interactive_viewer.ml:
```diff
@@ -60,9 +60,41 @@ let navigate z_patches (dir : direction) : unit =
+      (format_plural total_additions "addition" "additions")
+      (format_plural total_removals "removal" "removals")
+  in
+  W.string ~attr:Notty.A.(fg lightcyan) @@ Printf.sprintf "%s\n" operation_count
```
This `\n` is used to render an empty line. Is it intended ?

If it is, I'd prefer it not to be implemented like this. Instead an empty line should be explicitly added in the `W.vbox` with `W.string "";`.

In any case, this should be simplified to:
```suggestion
  W.string ~attr:Notty.A.(fg lightcyan) operation_count
```

maha-sachin replied:
@Julow, Thank you for your feedback! I'm glad to hear that you find the implementation satisfactory.
I'll address the comment regarding the empty line promptly. Once that's done, I'll let you know.

Julow replied:
I agree the output looks better without the newline.

I do not see any new modification in the PR, did you push to the same branch ?

maha-sachin replied:
@Julow wanted to check before making the change, the change should be in and ready to merge.

Julow reviewed (commented):
Looks very good !

I have a comment regarding the empty line. Otherwise, this is perfect and I would merge.

Julow reviewed (approved):
Perfect! Let's merge!

maha-sachin commented:
Thank you @Julow & @panglesd for merging the PR, @panglesd, for your detailed explanations and guidance throughout the process. Your insights were invaluable in understanding the concepts better. I appreciate your support and look forward to continuing to learn from @Julow and you.
